### PR TITLE
Convection diffusion vector CUDA

### PR DESCRIPTION
**Pull request by aneo-mderbane**

### CODE REVIEW

File: src/alge/cs_convection_diffusion.c [hunk 1]
```diff
@@ -4369,80 +4370,286 @@ cs_convection_diffusion_vector(int                         idtvar,
...
+      }
+    #endif
 
-# pragma omp parallel for
-  for (cs_lnum_t cell_id = 0; cell_id < n_cells_ext; cell_id++) {
-    for (int jsou = 0; jsou < 3; jsou++) {
-      for (int isou = 0; isou < 3; isou++)
-        grdpa[cell_id][isou][jsou] = 0.;
+    if(compute_cpu){
+      printf("convection Compute and tranferts time in us: CPU = %ld\n", elapsed);
```
Comment (ddiakiteaneo) [inline src/alge/cs_convection_diffusion.c]:
```suggestion
      printf("convection compute time in us: CPU = %ld\n", elapsed);
```

File: src/alge/cs_gradient.cxx [hunk 2]
```diff
@@ -6936,11 +6936,20 @@ _lsq_vector_gradient(const cs_mesh_t               *m,
...
-BFT_MALLOC(gradv_cpu, n_cells_ext, cs_real_33_t);
+  // Pour l'instant ces lignes sont pour moi
+  // Elles seront à enlever
+  compute_cuda  = false;
+  compute_cpu   = true;
+  res_cpu       = true;
+
+  // A ne pas garder dans la version finale
+  perf        = false;
+  accuracy    = false;
```
Comment (ddiakiteaneo) [inline src/alge/cs_gradient.cxx]: You can remove these lines or set all to false.

File: src/alge/cs_gradient_cuda.cuh [hunk 3]
```diff
@@ -96,6 +96,21 @@ cs_math_3_dot_product_cuda(const cs_real_t  u[3],
...
+
+  if(c_idx >= n_b_faces){
+    return;
+  }
+
+  cs_lnum_t f_id = c_idx / 3;
+  size_t i = c_idx % 3;
+  
+  _bc_coeff_b[f_id][i][i] = 1;
+}
```
Comment (ddiakiteaneo) [inline src/alge/cs_gradient_cuda.cuh:158]: Is this used anymore ? If not, remove them

File: src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    f_id = cell_i_faces[index];
+    face_sgn = cell_i_faces_sgn[index];
+
+    for (int jsou = 0; jsou < 3; jsou++) {
+      difv[jsou] = i_face_cog[f_id][jsou] - cell_cen[c_id1][jsou];
+      djfv[jsou] = i_face_cog[f_id][jsou] - cell_cen[c_id2][jsou];
+    }
+
+    /* x-y-z component, p = u, v, w */
+
```
Comment (ddiakiteaneo) [inline src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh]:
```suggestion
```
Remove comment in kernels

File: src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        pjf = pjf + grad[c_id2][isou][jsou] * djfv[jsou];
+      }
+
+      pfac = pjf;
+      if (i_massflux[f_id] * face_sgn > 0.) 
+        pfac = pif;
+      pfac *= face_sgn;
+
+      /* U gradient */
+
```
Comment (ddiakiteaneo) [inline src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh]:
```suggestion
```
Same remark

File: src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  cs_real_t diipbv[3];
+  cs_lnum_t f_id;
+  cs_lnum_t s_id = cell_b_faces_idx[c_id];
+  cs_lnum_t e_id = cell_b_faces_idx[c_id + 1];
+
+  for(cs_lnum_t index = s_id; index < e_id; index++){
+    f_id = cell_b_faces[index];
+
+    /* x-y-z components, p = u, v, w */
+
```
Comment (ddiakiteaneo) [inline src/alge/cs_slope_test_gradient_vector_cuda_gather.cuh]:
```suggestion
```
Same remark

File: src/alge/cs_convection_diffusion_cuda.cu [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+    if (alloc_mode == CS_ALLOC_HOST_DEVICE_SHARED)
+      cudaMemPrefetchAsync(val_h, size, device_id, stream);
+    else
+      cs_sync_h2d(val_h);
+  }
+
+  *val_d = _val_d;
+  *buf_d = _buf_d;
+}
```
Comment (ddiakiteaneo) [inline src/alge/cs_convection_diffusion_cuda.cu]:
```suggestion
```
To be removed !

File: src/alge/cs_gradient_cuda.cuh (hunk 3 above)
Comment (aneo-mderbane) [reply src/alge/cs_gradient_cuda.cuh:158]: Yes it's used

**Review by ddiakiteaneo — APPROVED** [review]

**Pull request merged**